### ISSUE TITLE
Image is stored to the card only after second run

### ISSUE THREAD

**Issue by MarekSalat** [googlesamples/android-Camera2Raw#4]
Hello, I am writing a simple application and I have used this example as scaffold. What I have found out that image is stored into the card after second run. When I call take picture method, image is not saved and it is saved only when I called the method twice or more. Here is why (I hope that the flow will be understandable)

When you call takePicture method plenty of code is executed but we are interested only to onCaptureCompleted method. 

The code creates completion task, after that result is set to the builder. But the completion task is created only if result is already set to the mCaptureResult. Here is the problem. setResult should be called before handleCompletionLocked or otherwise. 

``` java
@Override
        public void onCaptureCompleted(/*...*/) {
//...
            synchronized (mCameraStateLock) {
                // If we have all the results necessary, save the image to a file in the background.
                handleCompletionLocked(requestId, jpegBuilder, mJpegResultQueue); // creates task only if mCaptureResult != null

                if (jpegBuilder != null) {
                    jpegBuilder.setResult(result); // but here is code which sets mCaptureResult
// ...
                }

        }
```

Here is why. Task is created only if saver is not null.

``` java
    private void handleCompletionLocked(/*...*/) {
        if (builder == null) return;
        ImageSaver saver = builder.buildIfComplete();
        if (saver != null) {
            queue.remove(requestId);
            AsyncTask.THREAD_POOL_EXECUTOR.execute(saver);
        }
    }
```

because it is first run mCamptureResult is null thus buildIfComplete must return null

``` java
public synchronized ImageSaver buildIfComplete() {
                if (!isComplete()) { // mImage != null && mFile != null && mCaptureResult != null && mCharacteristics != null
                    return null;
                }
                return new ImageSaver(mImage, mFile, mCaptureResult, mCharacteristics, mContext,
                        mReader);
            }
```

more over mCaptureResult is set only in one place and this is setResult method.

The fix is simple just move setResult before handleCompletionLocked or move handleCompletionLocked after setResult.

Please note that I am not an android master :) 

Also I have a problem with rotation of the saved image on my nexus 7. Image is always flipped upside down but that is another issue I suppose.

**Comment by m0nkiii** [comment]
You are right! I had the same issue and I figured out the same fix and now it works every time.

**Comment by MattiasSvenJonsson** [comment]
You are right!
I had the same problem and I figured out the same fix as you.

**Comment by perracodex** [comment]
I’ve found that the next callbacks are not triggered in the same order under different API levels. 

1) ImageReader.OnImageAvailableListener.onImageAvailable()
2) CameraCaptureSession.CaptureCallback.onCaptureCompleted()

For example, in the Nexus 6 (API 6.0.1) follows the above calling order, where the Moto X 1st gen (API 5.1) the callbacks are triggered in reverse. I’m assuming is at API level and not a per device issue, though further testing would be required.

And your fix is correct. The builder.setResult() should be above the handleCompletionLocked() method, as you can see in the dequeueAndSaveImage() method, where the builder.setImage() is also set before handleCompletionLocked(). This will ensure that no matter the order of the callbacks, once all the Builder required fields are set, from whichever callback order is followed, the task will be finally executed.

**Comment by codingjeremy** [comment]
This sample has been migrated to a new location (check README for more information). 

[As recommended by GitHub][1], we are closing all issues and pull requests now that this older repo will be archived. 

If you still see this issue in the updated repo, please reopen the issue/PR there. Thank you!

[1]: https://help.github.com/en/articles/archiving-repositories
